### PR TITLE
Avivash/91 add recovery kit

### PR DESCRIPTION
**Pull request by avivash**
# Description

- Add recovery kit download to account settings page
- Update registration form UI
- Add recovery kit upload flow
- ~~I've also added the UI for when a user doesn't have a recovery kit, but it's commented out because we need to discuss how we want that flow to work. It may make more sense for me to move that into its own branch~~ I ended up [moving the no recovery kit code to its own branch](https://github.com/webnative-examples/webnative-app-template/tree/avivash/91-no-recovery-kit-flow))

**Note: now that the recovery kit flow has been implemented in this app. I think a good fast-follow issue could be to add some of that could to webnative directly, as per Steven's suggestion 👍🏼** 

## Link to issue

https://github.com/webnative-examples/webnative-app-template/issues/91

## Type of change

- [x] New feature (non-breaking change that adds functionality)
- [x] Refactor (non-breaking change that updates existing functionality)

## Screenshots/Screencaps

**Recovery kit download -> upload flow(happy path):**
https://www.loom.com/share/0b123537214449259a2046e8204b6c61

**Recovery kit upload error state:**
image

**No recovery kit UI:**
image

**New registration form UI:**
https://www.loom.com/share/3c31733a8ad541fa8affccc37ad5e8e2

### CODE REVIEW

**Comment by avivash** [comment]
It seems i'm not a member of the `fission-codes` vercel org, so I can't see the deploy logs to isolate the error that's happening. I need to verify my account with my phone number, so I'm planning on turning my roaming back on tomorrow so I can use my phone down here because I need to do a couple different things with it, but am trying to limit the amount of times I have to pay $15 per day 😅 

Any chance anyone else can screenshot the error and send it to me in the meantime, please?

**Comment by avivash** [comment]
Also, keep in mind my internet connection is quite slow, so I think the loading times will be faster than what those loom videos show 🙈

File: src/lib/account-settings.ts [hunk 1]
```diff
@@ -181,3 +183,70 @@ export const uploadAvatarToWNFS = async (image: File): Promise<void> => {
...
+      components: { crypto, reference, storage }
+    },
+    username: {
+      full,
+      trimmed
+    }
+  } = getStore(sessionStore)
+
+  // Get the user's read-key and base64 encode it
+  const accountDID = await rootDID({ crypto, reference, storage })
```
Comment (icidasset) [inline src/lib/account-settings.ts:199]: I would avoid using functions directly from component implementations.
I would use the following instead:
```ts
await reference.didRoot.lookup(username)
```
We could add a shorthand function for this to the program if that's easier.
Like we just did with `program.agentDID()`

---

Also, I'm thinking of removing that `rootDID` function from the wnfs-auth implementation.
It looks up the root DID via UCAN proofs, which we should probably avoid.
For two reasons:
1. This lookup should be consistent with the other lookups in Webnative. The other day I had a situation where the root DID from the UCAN didn't match the one in DNS, so I couldn't load my file system.
2. When we adopt UCAN 0.9, we won't have nested UCANs in the current sense anymore, they'll be CIDs or another type of pointer pointing at the other UCAN. So this recursive lookup might not be as easy anymore.

Thoughts? cc @bgins

Comment (avivash) [reply src/lib/account-settings.ts:199]: sounds good! i'll use `await reference.didRoot.lookup(username)` for now :ok_hand:

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+          parts[1].replace(/(\r\n|\n|\r)/gm, ''),
+          'base64pad'
+        )
+
+        storage.setItem(USERNAME_STORAGE_KEY, username)
+
+        const rootDID = await reference.didRoot.lookup(hashedUsername)
+
+        // Store the accountDID and readKey in webnative so they can be used internally load the file system
+        await store({
```
[the remark this answers is not in the document]
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte]: updated!

File: src/lib/account-settings.ts (hunk 1 above)
Comment (avivash) [reply src/lib/account-settings.ts]: updated 👍🏼 maybe we can look at adding the shorthand function to webnative in an upcoming release?

Comment (icidasset) [reply src/lib/account-settings.ts]: Sounds good 👍

Comment (icidasset) [reply src/lib/account-settings.ts]: Continueing root DID discussion here: https://github.com/fission-codes/webnative/pull/435/files#r1047523803

**Comment by avivash** [comment]
the vercel gods have been appeased 🙇🏼

File: src/lib/account-settings.ts (hunk 1 above)
Comment (bgins) [reply src/lib/account-settings.ts]: > Also, I'm thinking of removing that rootDID function from the wnfs-auth implementation.
It looks up the root DID via UCAN proofs, which we should probably avoid.
For two reasons:
> 1. This lookup should be consistent with the other lookups in Webnative. The other day I had a situation where the root DID from the UCAN didn't match the one in DNS, so I couldn't load my file system.
> 2. When we adopt UCAN 0.9, we won't have nested UCANs in the current sense anymore, they'll be CIDs or another type of pointer pointing at the other UCAN. So this recursive lookup might not be as easy anymore.

> Thoughts? cc @bgins

A bit late to the party on this, but I'm agreed that we probably won't want to rely on finding the root DID from the UCAN in the future. 👌

File: src/lib/init.ts [hunk 3]
```diff
@@ -5,12 +5,14 @@ import { filesystemStore, sessionStore } from '../stores'
 import { getBackupStatus, type BackupStatus } from '$lib/auth/backup'
 import { USERNAME_STORAGE_KEY } from '$lib/auth/account'
 
+export const NAMESPACE = { creator: 'Fission', name: 'WAT' }
```
Comment (bgins) [inline src/lib/init.ts:8]: It might be worth moving this to `src/lib/app-info.ts` to make it easier for folks to change in their apps. Also worth mentioning it in the README in the "Rename your application" section.

**Review by bgins — COMMENTED** [review]
The recovery flow and the recovery kit file are looking great! ✨ 

A couple of issues. The recovery isn't working reliably. In some cases, when a recovery kit is uploaded, the app hangs at "Processing recovery kit". I'm not seeing any errors in the console. I'd be happy to run through a failing case if it would help to debug.

After a successful recovery, there was an issue with account linking. On starting the account linking producer, there was this error in the console:

linking-issue

Maybe something with how the recovered username is stored? Not seeing this before the account is recovered.

File: src/lib/init.ts (hunk 3 above)
Comment (avivash) [reply src/lib/init.ts:8]: good call!

Comment (avivash) [reply src/lib/init.ts]: done!

**Comment by avivash** [comment]
@bgins I've updated the `HasRecoveryKit` flow to create a new user now and device linking seems to be working smoothly post-recovery 🎉

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+          parts[1].replace(/(\r\n|\n|\r)/gm, ''),
+          'base64pad'
+        )
+
+        const oldUsername = parts[0].replace(/(\r\n|\n|\r)/gm, '')
+        const hashedOldUsername = await prepareUsername(oldUsername)
+        const oldRootDID = await reference.didRoot.lookup(hashedOldUsername)
+
+        // Construct a new username using the old `trimmed` name and `oldRootDID`
+        const newUsername = `${oldUsername.split('#')[0]}#${oldRootDID}`
```
Comment (icidasset) [inline src/components/auth/recover/HasRecoveryKit.svelte:56]: Shouldn't this have the new root DID instead?

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+          dependencies: $sessionStore.program.components,
+          potency: 'APPEND',
+          resource: '*',
+          proof: proof ? proof : undefined,
+          lifetimeInSeconds: 60 * 3, // Three minutes
+          audience: issuer,
+          issuer
+        })
+
+        const newRootDID = await reference.didRoot.lookup(hashedNewUsername)
```
Comment (icidasset) [inline src/components/auth/recover/HasRecoveryKit.svelte:84]: You can skip this. This'll be the local DID, ie. the `issuer` above. You'll also need this to construct the new root DID for the username.

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        const oldRootCID = await reference.dataRoot.lookup(hashedOldUsername)
+
+        // Update the dataRoot of the new user
+        await reference.dataRoot.update(oldRootCID, ucan)
+
+        // Store the accountDID and readKey in webnative so they can be used internally load the file system
```
Comment (icidasset) [inline src/components/auth/recover/HasRecoveryKit.svelte:90]: I would rephrase this a bit. The account DID is not stored to be used elsewhere, the only purpose of passing it in here is to namespace the read key (in other words, only used in the storage property name).

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 4 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte:56]: I was wondering about this. Am I able to get the new root DID if I haven't registered the new user yet, though?

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 6 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte:90]: ahhh I see, thanks 👍🏼

Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte]: updated

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 4 above)
Comment (icidasset) [reply src/components/auth/recover/HasRecoveryKit.svelte:56]: Yup, this'll be the local DID, since you're registering an account here. See my other comment about `issuer`.

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        // Register the user with the `hashedNewUsername`
+        const { success } = await authStrategy.register({
+          username: hashedNewUsername
+        })
+        if (!success) {
+          throw new Error('Failed to register new user')
+        }
+
+        // Build an ephemeral UCAN to allow the
+        const issuer = await DID.write(crypto)
```
Comment (icidasset) [inline src/components/auth/recover/HasRecoveryKit.svelte:70]: You can also say `program.agentDID()` here, whatever you prefer.

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 5 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte]: done

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 4 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte]: ahh yes, thanks! updated 👍🏼

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 7 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte]: done 👍🏼

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            components: { crypto, reference, storage }
+          }
+        } = $sessionStore
+
+        const parts = (event.target.result as string)
+          .split('username: ')[1]
+          .split('key: ')
+
+        const readKey = uint8arrays.fromString(
+          parts[1].replace(/(\r\n|\n|\r)/gm, ''),
```
Comment (bgins) [inline src/components/auth/recover/HasRecoveryKit.svelte:47]: It might be good to add a comment explaining this regexp. They are always hard to read without some context.

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+          'base64pad'
+        )
+
+        const oldUsername = parts[0].replace(/(\r\n|\n|\r)/gm, '')
```
Comment (bgins) [inline src/components/auth/recover/HasRecoveryKit.svelte:51]: A comment on the regexp might be good here as well.

File: src/components/auth/recover/HasRecoveryKit.svelte [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+        // Register the user with the `hashedNewUsername`
+        const { success } = await authStrategy.register({
+          username: hashedNewUsername
+        })
+        if (!success) {
+          throw new Error('Failed to register new user')
+        }
+
+        // Build an ephemeral UCAN to allow the
```
Comment (bgins) [inline src/components/auth/recover/HasRecoveryKit.svelte:69]: Looks like this comment got cut off?

File: src/components/auth/recover/RecoveryKitButton.svelte [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  $: if (files) {
+    handleFileInput(files)
+  }
+
+  $: buttonData = {
+    [RECOVERY_STATES.Processing]: {
+      text: 'Processing recovery kit...',
+      props: {
+        disabled: state === RECOVERY_STATES.Processing,
+        $$on_click: () => {}
```
Comment (bgins) [inline src/components/auth/recover/RecoveryKitButton.svelte:22]: Ah interesting! What does the `$$` do?

**Review by bgins — APPROVED** [review]
Looking good! 🎉 

Left a few comments with some suggestions and requests for additional comments.

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 10 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte]: updated!

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 9 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte:53]: done!

File: src/components/auth/recover/HasRecoveryKit.svelte (hunk 8 above)
Comment (avivash) [reply src/components/auth/recover/HasRecoveryKit.svelte:48]: done!

File: README.md [hunk 12]
```diff
@@ -81,6 +81,7 @@ The app template is designed to be easy for you to _make it your own._ Here's ho
 
     - Change `appName` to the name of your app.
     - Change `appDescription` to a simple, 1-sentence description of your app.
+    - Change `webnativeNamespace` to a represent your specific project details.
```
[the remark this answers is not in the document]
Comment (avivash) [reply README.md]: done!

File: src/components/auth/recover/RecoveryKitButton.svelte (hunk 11 above)
Comment (avivash) [reply src/components/auth/recover/RecoveryKitButton.svelte:22]: Nothing in particular 😅 i'm just using it here to denote that this is not a spreadable prop. I've seen it done in some other svelte projects 🤷🏼
